### PR TITLE
Don't reconcile Works that are being deleted; set workplacement finalizers after writing files

### LINKED ISSUE #203: flakey system test
https://app.circleci.com/pipelines/github/syntasso/kratix/2122/workflows/d3be1b25-8cbb-48e4-b51b-4ae8083bde3f/jobs/11711

https://app.circleci.com/pipelines/github/syntasso/kratix/2085/workflows/e930bcae-1e59-43b7-8b40-c50516eecc65/jobs/11551

failing every 5/6 runs

### PR DESCRIPTION
**Pull request by aclevername**
closes #203 

green runs (the red is a different testing flaking, I'm investigating and will address in a different PR) 
Screenshot 2024-07-24 at 14 11 39

this bug was pretty nasty to debug, and highlighted two issues:
- Destination controller triggers the reconciliation of ALL works, even if the work in question is being deleted, which causees it to recreate a workplacement immediately after its been deleted (race conditon)
- A workplacement gets the finalizer added for deleting the files it creates *before* it creates them, which means if a workplacement is created, finalizers added (requeue triggered), workplacement deleted, requeue starts the deletion path despite no files having been created.

## Example failure

Chronological order:

1. An existing healthy work get deleted, triggers deletion of workplacement
1  Work being deleted

2. Workplacement gets successfully deleted
2  successfull deletion of wp

3. The destination controller then triggers a reconcilation of all works (via the code directly, not via the work controller), the Work placement which is being deleted hasn't been deleted *yet*, because there is a time delay between the workplacement successfully being deleted, and the work resource having its finalzier removed. This reconciliation therefore then creates a new workplacement for the work
3  destination triggering reconciliation of deleting work resource

4. Workplacement gets created with finalizers, and then requeues (no files written yet!)
4  workplacement recreated, adding finalizers
 Work being deleted](https://github.com/user-attachments/assets/bcad7c45-086e-4c10-82db-bb120a8a552f)
 
5. The workplacement gets marked as deleted, even though its not created any files yet
5  deletion triggered- before workplacement actualy writes files

6. Workplacement deletion fails because the finalizers exist, but the files do not
6  workplacement deletion fails because it never wrote the fails, but has the finalizers

### CODE REVIEW

File: controllers/workplacement_controller.go [hunk 1]
```diff
@@ -120,15 +119,20 @@ func (r *WorkPlacementReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
 			return ctrl.Result{}, err
 		}
 	}
-	logger.Info("WorkPlacement successfully reconciled", "workPlacement", workPlacement.Name, "versionID", versionID)
 
+	if missingFinalizers := checkWorkPlacementFinalizers(workPlacement, filepathMode); len(missingFinalizers) > 0 {
+		return addFinalizers(opts, workPlacement, missingFinalizers)
+	}
+
+	logger.Info("WorkPlacement successfully reconciled", "workPlacement", workPlacement.Name, "versionID", versionID)
```
Comment (aclevername) [inline controllers/workplacement_controller.go:135]: the order is now:

1. update the files in statestore
1. update the status of the workplacement with the latest VersionID
1. update the finalizers

I considered swapping 2/3 round, i dont mind tbh. They both can't be done at the same time annoyingly since its an update to the status and the metadata, which is two separate API calls, which won't work without requeuing since the resource will have changed between the two API calls

File: lib/writers/git.go [hunk 2]
```diff
@@ -219,7 +219,7 @@ func (g *GitWriter) ReadFile(filePath string) ([]byte, error) {
 	defer os.RemoveAll(filepath.Dir(localTmpDir))
 
 	if _, err := worktree.Filesystem.Lstat(fullPath); err != nil {
-		logger.Error(err, "could not stat file")
+		logger.Info("could not stat file", "err", err)
```
Comment (aclevername) [inline lib/writers/git.go:222]: sometimes this is fine, so having this error output was making me think the reconciler exited when it hasn't. the cosnume of this detects if it should error log or ignore

File: controllers/workplacement_controller_test.go [hunk 3]
```diff
@@ -165,7 +165,7 @@ var _ = Describe("WorkplacementReconciler", func() {
 				Expect(result).To(Equal(ctrl.Result{}))
 
 				By("calling UpdateFiles()")
-				Expect(fakeWriter.UpdateFilesCallCount()).To(Equal(1))
+				Expect(fakeWriter.UpdateFilesCallCount()).To(Equal(2))
```
Comment (ChunyiLyu) [inline controllers/workplacement_controller_test.go:168]: Why is this needed?

File: controllers/work_controller.go [hunk 4]
```diff
@@ -124,5 +127,30 @@ func (r *WorkReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+
+func (r *WorkReconciler) requestReconilationOfAllWorksOnDestinationCreateOrUpdate(ctx context.Context, obj client.Object) []reconcile.Request {
+	dest := obj.(*v1alpha1.Destination)
+
+	if dest.GetDeletionTimestamp() != nil {
+		return nil
+	}
+
+	allWorks := &v1alpha1.WorkList{}
+	err := r.Client.List(ctx, allWorks)
```
Comment (ChunyiLyu) [inline controllers/work_controller.go:145]: `scheduler.ReconcileAllDependencyWorks()` only reconciles if `work.IsDependency()` is true. Should we have the same check in here?

File: controllers/workplacement_controller_test.go (hunk 3 above)
Comment (aclevername) [reply controllers/workplacement_controller_test.go:168]: the reconciliation loop runs twice, because the finalizer is now after the update an additional call is made to the update. Before it was finazlers->requque->finalizers->update->done, now its update->requeue->update->finalizers->done

File: controllers/work_controller.go (hunk 4 above)
Comment (aclevername) [reply controllers/work_controller.go:145]: I don't think so. If a work is for a resource request and is not scheduled (no matching destination) we do want to reconcile the work. Also if a Work for a resource request is scheduled to a destination, and then that destination is updated to no longer match the labels that made it get scheduled to by the work to begin with we update the work status.

**Review by ChunyiLyu — APPROVED** [review]
